txpool/txpool.go

```diff
@@ -288,89 +302,128 @@ func (t *TxPool) validateTx(tx *types.Transaction) error {
 	return nil
 }
 
-type txQueue struct {
-	txs       txHeap
+// txHeapWrapper is a wrapper object for account based transactions
+type txHeapWrapper struct {
+	// txs is the actual min heap (nonce ordered) for account transactions
+	txs txHeap
+
+	// nextNonce is a field indicating what should be the next
+	// valid nonce for the account transaction
 	nextNonce uint64
 }
 
-func newTxQueue() *txQueue {
-	return &txQueue{
+// newTxHeapWrapper creates a new account based tx heap
+func newTxHeapWrapper() *txHeapWrapper {
+	return &txHeapWrapper{
 		txs: txHeap{},
 	}
 }
 
-func (t *txQueue) Reset(txns ...*types.Transaction) {
-	lowestNonce := t.nextNonce
-	for _, txn := range txns {
-		if txn.Nonce < lowestNonce {
-			lowestNonce = txn.Nonce
-		}
-		t.Push(txn)
-	}
-	t.nextNonce = lowestNonce
-}
-
-// Add adds a new tx into the queue
-func (t *txQueue) Add(tx *types.Transaction) {
+// Add adds a new tx onto the account based tx heap
+func (t *txHeapWrapper) Add(tx *types.Transaction) {
 	t.Push(tx)
 }
 
-// Promote promotes all the new valid transactions
-func (t *txQueue) Promote() []*types.Transaction {
-	// Remove elements lower than nonce
+// pruneLowNonceTx removes any transactions from the account tx queue
+// that have a lower nonce than the current account nonce in state
+func (t *txHeapWrapper) pruneLowNonceTx() {
 	for {
+		// Grab the min-nonce transaction from the heap
 		tx := t.Peek()
 		if tx == nil || tx.Nonce >= t.nextNonce {
 			break
 		}
+
+		// Drop it from the heap
 		t.Pop()
 	}
+}
+
+// Promote promotes all the new valid transactions
+func (t *txHeapWrapper) Promote() []*types.Transaction {
+	// Remove elements lower than nonce
+	t.pruneLowNonceTx()
 
 	// Promote elements
 	tx := t.Peek()
 	if tx == nil || tx.Nonce != t.nextNonce {
+		// Nothing to promote
 		return nil
 	}
 
 	promote := []*types.Transaction{}
+	higherNonceTxs := []*types.Transaction{}
+
+	reinsertFunc := func() {
+		// Reinsert the tx back to the account specific transaction queue
+		for _, highNonceTx := range higherNonceTxs {
+			t.Push(highNonceTx)
+		}
+	}
+
 	for {
 		promote = append(promote, tx)
 		t.Pop()
 
-		tx2 := t.Peek()
-		if tx2 == nil || tx.Nonce+1 != tx2.Nonce {
+		var nextTx *types.Transaction
+		if nextTx = t.Peek(); nextTx == nil {
+			break
+		}
+
+		if tx.Nonce+1 != nextTx.Nonce {
+			// Tx that have a higher nonce are shelved for later
+			// when they can actually be parsed
+			higherNonceTxs = append(higherNonceTxs, nextTx)
 			break
 		}
-		tx = tx2
+
+		tx = nextTx
 	}
+
 	if len(promote) == 0 {
+		// Nothing to promote
+		reinsertFunc()
+
 		return nil
 	}
```

Comment:
Can this `if` even happen?
inside of the `for` loop on the first line we append the `promote` slice with the tx, which can't be `nil` because we checked that on L349

Reply:
I think you're actually right.

Promoted will never be empty after the for loop, since we'll always be adding at least 1 element here:
https://github.com/0xPolygon/polygon-sdk/blob/a6edca6741245aa54f1210e56556dcc136d0bd1c/txpool/txpool.go#L365

Good catch 💯 

Removing now 👍 

30caf6e